README
marc is a ruby library for reading and writing MAchine Readable Cataloging
(MARC). More information about MARC can be found at <http://www.loc.gov/marc>.

USAGE 

  require 'marc'

  # reading records from a batch file
  reader = MARC::Reader.new('marc.dat')
  for record in reader
    # print out field 245 subfield a
    puts record['245']['a']
  end

  # creating a record 
  record = MARC::Record.new()
  record.append(MARC::DataField.new('100', '0',  ' ', ['a', 'John Doe']))

  # writing a record
  writer = MARC::Writer.new('marc.dat')
  writer.write(record)
  writer.close()

  # writing a record as XML
  writer = MARC::XMLWriter.new('marc.xml')
  writer.write(record)
  writer.close()
  
  # encoding a record
  MARC::Writer.encode(record) # or record.to_marc

INSTALLATION

  gem install marc

Or if you're using bundler, add to your Gemfile

  gem 'marc'
  
ERRATA

Source code at: https://github.com/ruby-marc/ruby-marc/

Find generated API docs at: http://rubydoc.info/gems/marc/frames

Run automated tests in source with `rake test`. 

Developers, release new version of gem to rubygems with `rake release` 
(bundler-supplied task). Note that one nice thing this will do is automatically
tag the version in git, very important for later figuring out what's going on.

Please send bugs, requests and comments to Code4Lib Mailing list (https://listserv.nd.edu/cgi-bin/wa?A0=CODE4LIB). 

AUTHORS

Kevin Clarke <ksclarke@gmail.com>
Bill Dueber <bill@dueber.com>
William Groppe <will.groppe@gmail.com>
Ross Singer <rossfsinger@gmail.com>
Ed Summers <ehs@pobox.com>


Changes
v1.1.2 (next)
- Fix JRubySTAXReader to use :: syntax for constant access as required by
  JRuby 9.3+.
- Document JAVA_OPTS workaround for JRubySTAXReader module encapsulation
  issue under Java 9+.
- Deprecate JRubySTAXReader.

v1.1.1 June 2021
- Fix a regression when normalizing indicator values when serializing marcxml

v1.1.0 June 2021
 - Add support for additional valid subfield codes in marcxml

v1.0.2 July 2017
 - Now (correctly) throw an error if datafield string is the empty string
   (thanks to @bibliotechy)

v1.0.1 February 2016
- Non-user-facing change in implementation of FieldMap strictly for performance

v1.0.0 January 2015
- Mostly changes that deal with encoding, plus the plunge to a 1.0 release

v0.5.0 April 2012
- Extensive rewrite of MARC::Reader (ISO 2709 binary reader) to provide a
  fairly complete and consistent handing of char encoding issues in ruby 1.9.
  - This code is well covered by automated tests, but ends up complex, there
    may be bugs, please report them.
  - May not work properly under jruby with non-unicode source encodings.
  - Still can't handle Marc8 encoding.
  - May not have entirely backwards compatible behavior with regard to char
    encodings under ruby 1.9.x as previous 0.4.x versions. Test your code.
    In particular, previous versions may have automatically _transcoded_
    non-unicode encodings to UTF-8 for you. This version will not do
    so unless you ask it to with correct arguments.

v0.4.4 Sat Mar 03 14:55:00 EDT 2012
- Fixed performance regression: strict reader will parse about 5x faster now
- Updated CHANGES file for first time in a long time :-)

v0.3.0 Wed Sep 23 21:51:00 EDT 2009
- Nokogiri and jrexml parser integration added as well as Ruby 1.9 support

v0.2.2 Tue Dec 30 09:50:33 EST 2008
- DataField tags that are all numeric are now padded with leading zeros

v0.2.1 Mon Aug 18 14:14:16 EDT 2008
- can now process records that have fields tags that are non-numeric (thanks
  Ross Singer)

v0.2.0 Wed Jun 11 12:42:20 EDT 2008
- added newline to output generated by REXML::Formatters::Default to make
  it a bit more friendly. REXML::Formatters::Pretty and Transitive just
  don't do what I want (whitespace in weird places).

v0.1.9 Thu Jun  5 12:00:01 EDT 2008
- small docfix change in XMLReader
- use REXML::Formatters::Default instead of deprecated REXML::Element.write

v0.1.8 Tue Nov 13 22:51:03 EST 2007
- added examples directory
- fixed problem with leading whitespace and the leader in xml reader
  (thanks Morgan Cundiff)

v0.1.7 Mon Nov 12 09:33:57 EST 2007
- updated Record.to_marc documentation to be a bit more precise
- removed doc references to MARC::Field which is no longer around
- changed from Artistic to MIT License

v0.1.6 Fri May  4 12:37:33 EDT 2007
- fixed bad record length test
- removed MARC::XMLWriter convert_to_utf8 which wasn't really working and
  shouldn't be there if it isn't good
- added unescaping of entities to MARC::XMLReader

v0.1.5 Tue May  1 16:50:02 EDT 2007
- docfix in MARC::DataField (thanks Jason Ronallo)
- multiple docfixes (thanks Jonathan Rochkind)

v0.1.4 Tue Jan  2 15:45:53 EST 2007
- fixed bug in MARC::XMLWriter that was outputting all control field tags as 00z
  (thanks Ross Singer)
- added :include_namespace option to MARC::XMLWriter::encode to include the
  marcxml namespace, which allows MARC::Record::to_xml to emit the namespace
  for a single record.

v0.1.3  Tue Jan  2 12:56:36 EST 2007
- added ability to map a MARC record to the Dublin Core fields.  Calling
  to_dublin_core on a MARC::Record returns a hash that has Dublin Core fields
  as the hash keys.

v0.1.2  Thu Dec 21 18:46:01 EST 2007
- fixed MARC::Record::to_xml so that it actually is tested and works (thanks
  Ross Singer)

v0.1.1
- added ability to pass File like objects to the constructor for
  MARC::XMLReader like MARC::Reader (thanks Jake Glenn)

v0.1.0  Wed Dec  6 15:40:40 EST 2006
- fixed pretty xml when stylesheet is used
- added value() to MARC::DataField
- added Rakefile for testing/building

v0.0.9  Tue Mar 28 10:02:16 CST 2006
- changed XMLWriter.write to output pretty-printed XML
- normalized Text in XML output
- added XMLWriter checks and replacements for bad subfield codes and indicator
  values
- added XMLWriter check and replacement for invalid control codes in xml data
  values
- added XMLWriter checks for values in the leader that are invalid MARCXML
- added bin/marc2xml
- collapsed tc_xmlreader.rb tc_xmlwriter.rb into tc_xml.rb for full write/read
  test.
- added :stylesheet argument to XLMWriter.new

v0.0.8  Mon Jan 16 22:31:00 EST 2006
- removed control tests out of tc_field.rb into tc_control.rb
- fixed some formatting
- changed control/field to controlfield/datafield
- added == check for controlfield
- removed namespace declarations on record elements in favor of default
  namespace on collection element
- added spaces around subfield code and delimeter in to_s
- fixed up relevant tests that were expecting old formatting
- fixed xmlreader strip_ns which was rerturning Nil when no namespace
  was found on an element (exposed by namespace changes).

v0.0.7  Mon Jan  2 21:39:28 CST 2006
- MARC::XMLWriter added
- removed encode/decode methods in MARC::MARC21 into MARC::Writer and
  MARC::Reader respectively. This required pushing MARC21 specific constants
  out into MARC::Constants which is required as necessary.
- moved encode from MARC::MARXML into MARC::XMLWriter and added constants
  to MARC::Constants
- added MARC::XMLReader for reading MARX as XML
- added xml reading tests
- fixed indentation to be two spaces

v0.0.6  Tue Oct 18 09:33:12 CDT 2005
- MARC::MARC21::decode throws an exception when a directory can't be found.
  Exception is caught and ignored in MARC::ForgivingReader

v0.0.5  Tue Oct 18 01:50:40 CDT 2005
- when unspecified field indicators are forced to blanks
- checking for when a field appears to not have indicators and subfields in
  which case the field is skipped entirely


lib/marc/record.rb
module MARC
  # The FieldMap is an Array of DataFields and Controlfields.
  # It also contains a Hash representation
  # of the fields for faster lookups (under certain conditions)
  class FieldMap < Array
    attr_reader :tags
    attr_accessor :clean

    def initialize
      @tags = {}
      @clean = true
    end

    # Rebuild the HashWithChecksumAttribute with the current
    # values of the fields Array
    def reindex
      @tags = {}
      each_with_index do |field, i|
        @tags[field.tag] ||= []
        @tags[field.tag] << i
      end
      @clean = true
    end

    # Returns an array of all of the tags that appear in the record (not in the order they appear, however).
    def tag_list
      reindex unless @clean
      @tags.keys
    end

    # Returns an array of fields, in the order they appear, according to their tag.
    # The tags argument can be a string (e.g. '245'), an array (['100','700','800'])
    # or a range (('600'..'699')).

    def each_by_tag(tags)
      reindex unless @clean
      indices = []
      # Get all the indices associated with the tags
      Array(tags).each do |t|
        indices.concat @tags[t] if @tags[t]
      end

      # Remove any nils
      indices.compact!
      return [] if indices.empty?

      # Sort it, so we get the fields back in the order they appear in the record
      indices.sort!

      indices.each do |tag|
        yield self[tag]
      end
    end

    # Freeze for immutability, first reindexing if needed.
    # A frozen FieldMap is safe for concurrent access, and also
    # can more easily avoid accidental reindexing on even read-only use.
    def freeze
      reindex unless @clean
      super
    end
  end

  # A class that represents an individual MARC record. Every record
  # is made up of a collection of MARC::DataField objects.
  #
  # MARC::Record mixes in Enumerable to enable access to constituent
  # DataFields. For example, to return a list of all subject DataFields:
  #
  #   record.find_all {|field| field.tag =~ /^6../}
  #
  # The accessor 'fields' is also an Array of MARC::DataField objects which
  # the client can modify if neccesary.
  #
  #   record.fields.delete(field)
  #
  # Other accessor attribute: 'leader' for record leader as String
  #
  # == High-performance lookup by tag
  #
  # A frequent use case is looking up fields in a MARC record by tag, such
  # as 'all the 500 fields'.  Certain methods can use a hash keyed by
  # tag name for higher performance lookup by tag.  The hash is lazily
  # created on first access -- there is some cost of creating the hash,
  # testing shows you get a performance advantage to using the hash-based
  # methods if you are doing at least a dozen lookups.
  #
  #     record.fields("500")  # returns an array
  #     record.each_by_tag("500") {|field| ... }
  #     record.fields(['100', '700'])   # can also use an array in both methods
  #     record.each_by_tag( 600..699 )  # or a range
  #
  # == Freezing for thread-safety and high performance
  #
  # MARC::Record is not generally safe for sharing between threads.
  # Even if you think you are just acccessing it read-only,
  # you may accidentally trigger a reindex of the by-tag cache (see above).
  #
  # However, after you are done constructing a Record, you can mark
  # the `fields` array as immutable. This makes a Record safe for sharing
  # between threads for read-only use, and also helps you avoid accidentally
  # triggering a reindex, as accidental reindexes can harm by-tag
  # lookup performance.
  #
  #     record.fields.freeze
  class Record
    include Enumerable

    # the record fields
    # attr_reader :fields

    # the record leader
    attr_accessor :leader

    def initialize
      @fields = FieldMap.new
      # leader is 24 bytes
      @leader = " " * 24
      # leader defaults:
      # http://www.loc.gov/marc/bibliographic/ecbdldrd.html
      @leader[10..11] = "22"
      @leader[20..23] = "4500"
    end

    # Returns true if there are no error messages associated with the record
    def valid?
      errors.none?
    end

    # Returns an array of validation errors for all fields in the record
    def errors
      @fields.flat_map(&:errors)
    end

    # add a field to the record
    #   record.append(MARC::DataField.new( '100', '2', '0', ['a', 'Fred']))

    def append(field)
      @fields.push(field)
      @fields.clean = false
    end

    # alias to append

    def <<(field)
      append(field)
    end

    # each() is here to support iterating and searching since MARC::Record
    # mixes in Enumerable
    #
    # iterating through the fields in a record:
    #   record.each { |f| print f }
    #
    # getting the 245
    #   title = record.find {|f| f.tag == '245'}
    #
    # getting all subjects
    #   subjects = record.find_all {|f| ('600'..'699') === f.tag}

    def each
      @fields.each do |field|
        yield field
      end
    end

    # A more convenient way to iterate over each field with a given tag.
    # The filter argument can be a string, array or range.
    def each_by_tag(filter)
      @fields.each_by_tag(filter) { |tag| yield tag }
    end

    # You can lookup fields using this shorthand:
    #   title = record['245']

    def [](tag)
      find { |f| f.tag == tag }
    end

    # Provides a backwards compatible means to access the FieldMap.
    # No argument returns the FieldMap array in entirety.  Providing
    # a string, array or range of tags will return an array of fields
    # in the order they appear in the record.
    def fields(filter = nil)
      unless filter
        # Since we're returning the FieldMap object, which the caller
        # may mutate, we precautionarily mark dirty -- unless it's frozen
        # immutable.
        @fields.clean = false unless @fields.frozen?
        return @fields
      end
      @fields.reindex unless @fields.clean
      flds = []
      if filter.is_a?(String) && @fields.tags[filter]
        @fields.tags[filter].each do |idx|
          flds << @fields[idx]
        end
      elsif filter.is_a?(Array) || filter.is_a?(Range)
        @fields.each_by_tag(filter) do |tag|
          flds << tag
        end
      end
      flds
    end

    # Returns an array of all of the tags that appear in the record (not necessarily in the order they appear).
    def tags
      @fields.tag_list
    end

    # Factory method for creating a MARC::Record from MARC21 in
    # transmission format.
    #
    #   record = MARC::Record.new_from_marc(marc21)
    #
    # in cases where you might be working with somewhat flawed
    # MARC data you may want to use the :forgiving parameter which
    # will bypass using field byte offsets and simply look for the
    # end of field byte to figure out the end of fields.
    #
    #  record = MARC::Record.new_from_marc(marc21, :forgiving => true)

    def self.new_from_marc(raw, params = {})
      MARC::Reader.decode(raw, params)
    end

    # Returns a record in MARC21 transmission format (ANSI Z39.2).
    # Really this is just a wrapper around MARC::MARC21::encode
    #
    #   marc = record.to_marc()

    def to_marc
      MARC::Writer.encode(self)
    end

    # Handy method for returning the MARCXML serialization for a
    # MARC::Record object. You'll get back a REXML::Document object.
    # Really this is just a wrapper around MARC::XMLWriter::encode
    #
    #   xml_doc = record.to_xml()

    def to_xml
      MARC::XMLWriter.encode(self, include_namespace: true)
    end

    # Create the actual XML string (as opposed to #to_xml which, for historic reasons,
    # returns an REXML document)
    # @param [Boolean] fast_but_unsafe Use the fast MARC::UnsafeXMLWriter code
    # @return [String] MARC-XML encoding of the record
    def to_xml_string(fast_but_unsafe: false)
      if fast_but_unsafe
        MARC::UnsafeXMLWriter.encode(self)
      else
        to_xml.to_s
      end
    end

    # Handy method for returning a hash mapping this records values
    # to the Dublin Core.
    #
    #   dc = record.to_dublin_core()
    #   print dc['title']

    def to_dublin_core
      MARC::DublinCore.map(self)
    end

    # Return a marc-hash version of the record
    def to_marchash
      {"type" => "marc-hash", "version" => [MARCHASH_MAJOR_VERSION, MARCHASH_MINOR_VERSION], "leader" => leader, "fields" => map { |f| f.to_marchash }}
    end

    # Factory method for creating a new MARC::Record from
    # a marchash object
    #
    # record = MARC::Record->new_from_marchash(mh)

    def self.new_from_marchash(mh)
      r = new
      r.leader = mh["leader"]
      mh["fields"].each do |f|
        if f.length == 2
          r << MARC::ControlField.new(f[0], f[1])
        elsif r << MARC::DataField.new(f[0], f[1], f[2], *f[3])
        end
      end
      r
    end

    # Returns a (roundtrippable) hash representation for MARC-in-JSON
    def to_hash
      record_hash = {"leader" => @leader, "fields" => []}
      @fields.each do |field|
        record_hash["fields"] << field.to_hash
      end
      record_hash
    end

    # Return an actual json-encoded string.
    def to_json_document
      MARC::JSONLWriter.encode(self)
    end

    def self.new_from_hash(h)
      r = new
      r.leader = h["leader"]
      h["fields"]&.each do |position|
        position.each_pair do |tag, field|
          if field.is_a?(Hash)
            f = MARC::DataField.new(tag, field["ind1"], field["ind2"])
            field["subfields"].each do |pos|
              pos.each_pair do |code, value|
                f.append MARC::Subfield.new(code, value)
              end
            end
            r << f
          else
            r << MARC::ControlField.new(tag, field)
          end
        end
      end
      r
    end

    # Returns a string version of the record, suitable for printing

    def to_s
      str = "LEADER #{leader}\n"
      each do |field|
        str += field.to_s + "\n"
      end
      str
    end

    # For testing if two records can be considered equal.

    def ==(other)
      to_s == other.to_s
    end

    # Handy for using a record in a regex:
    #   if record =~ /Gravity's Rainbow/ then print "Slothrop" end

    def =~(regex)
      to_s =~ regex
    end
  end
end
